### PR TITLE
PARQUET-1044: Use compression libraries from Apache Arrow

### PR DESCRIPTION
**Pull request by wesm**
Depends on https://github.com/apache/arrow/pull/771; will update Arrow version once that's merged

### CODE REVIEW

**Comment by wesm** [comment]
If @advancedxy, @majetideepak, or others could take a look when they get a chance, that would be great! Thanks

File: src/parquet/util/memory.h [hunk 1]
```diff
@@ -31,35 +31,38 @@
...
-    if (!_s.ok()) {                                 \
-      std::stringstream ss;                         \
-      ss << "Arrow error: " << _s.ToString();       \
-      ::parquet::ParquetException::Throw(ss.str()); \
-    }                                               \
-  } while (0);
-
 namespace parquet {
 
+static inline std::unique_ptr<::arrow::Codec> GetArrowCodec(Compression::type codec) {
```
Comment (advancedxy) [inline src/parquet/util/memory.h]: How about change to `GetCodecFromArrow` and reuse use codec_type?

ArrowCodec seems a bit weird..

Comment (wesm) [inline src/parquet/util/memory.h]: Sure, no problem

**Comment by wesm** [comment]
Sure, I'll back out the clang-format changes and get the build passing

File: benchmarks/decode_benchmark.cc [hunk 2]
```diff
@@ -277,19 +278,20 @@ uint64_t TestBinaryPackedEncoding(const char* name, const std::vector<int64_t>&
...
   const uint8_t* raw_data = reinterpret_cast<const uint8_t*>(&data[0]);
   int uncompressed_len = data.size() * sizeof(int64_t);
   uint8_t* decompressed_data = new uint8_t[uncompressed_len];
 
   int max_compressed_size = codec->MaxCompressedLen(uncompressed_len, raw_data);
   uint8_t* compressed_data = new uint8_t[max_compressed_size];
-  int compressed_len =
-      codec->Compress(uncompressed_len, raw_data, max_compressed_size, compressed_data);
+  int64_t compressed_len;
+  codec->Compress(
```
Comment (xhochy) [inline benchmarks/decode_benchmark.cc]: Unchecked status?

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-1150 about the compiler warning that showed up in AdaptiveIntBuilder. Interesting that it only happened in the debug build on MSVC

File: benchmarks/decode_benchmark.cc (hunk 2 above)
Comment (wesm) [inline benchmarks/decode_benchmark.cc]: Fixed

**Comment by wesm** [comment]
Here's the Appveyor build: https://ci.appveyor.com/project/wesm/parquet-cpp/build/1.0.43.

**Comment by majetideepak** [comment]
I am seeing parquet test linking issues after this change when statically linking with libarrow.
```
libarrow.a(compression.cc.o): In function `arrow::SnappyCodec::Decompress(long, unsigned char const*, long, unsigned char*)':
arrow/cpp/src/arrow/util/compression.cc:282: undefined reference to `snappy::RawUncompress(char const*, unsigned long, char*)'
libarrow.a(compression.cc.o): In function `arrow::SnappyCodec::MaxCompressedLen(long, unsigned char const*)':
arrow/cpp/src/arrow/util/compression.cc:289: undefined reference to `snappy::MaxCompressedLength(unsigned long)'
....................
....................
```
```
nm libarrow.a | grep snappy
                 U _ZN6snappy11RawCompressEPKcmPcPm
                 U _ZN6snappy13RawUncompressEPKcmPc
                 U _ZN6snappy19MaxCompressedLengthEm
```
How do I get the arrow link line to parquet?

**Comment by wesm** [comment]
Now that snappy is in the Arrow toolchain, if you're statically linking you'll have to also statically link libsnappy.a. I guess we should not support static linking in `libparquet.so` now

**Pull request closed without merging**